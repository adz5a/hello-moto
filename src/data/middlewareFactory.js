import isFunction from "lodash/isFunction";
import {
    ACTIONFACTORY,
    fromMiddleware
} from "data/commons";

const MIDDLEWARE = ACTIONFACTORY("__middleware__");
const PROCESSING = MIDDLEWARE("processing");
const ERROR = MIDDLEWARE("error");



export const tryCatch = (effect, wrapper) => {

    try {
        effect();
    } catch ( e ) {

        wrapper.value = e;
        return wrapper;

    };

};

export const process = store => ( type, meta ) => promise => promise
    .then(data => ({
        type,
        meta,
        data
    }))
    .catch( error => ({
        type: ERROR,
        data: {
            type,
            data: error
        },
        meta
    }));

export function MiddlewareFactory ( effectsMap ) {

    if ( typeof effectsMap !== "object" || effectsMap === null ) {

        throw new TypeError("Expects an object :: { key: function }");

    }

    const invalidKeys = Object.keys(effectsMap)
        .filter(key => !isFunction(effectsMap[key]));

    if ( invalidKeys.length > 0 ) {

        throw new TypeError("All keys must be functions, instead received non functions for : " + invalidKeys.join(" "));

    }


    const effects = new Map(
        Object.keys(effectsMap).map( key => [ key, effectsMap[key] ])
    );


    const errorWrapper = {};
    return function middleware ( store ) {

        const _process = process(store);
        return next => action => {

            const {Â type, meta = {} } = action;
            if ( effects.has(type) ) {

                const value = tryCatch(
                    () => effects.get(type)(action.data),
                    errorWrapper
                );

                const promise = value === errorWrapper ? 
                    Promise.reject(value.value) :
                    Promise.resolve(value)

                promise
                    .then(data => ({
                        type,
                        meta: fromMiddleware({
                            ...meta
                        }),
                        data
                    }))
                    .catch( error => ({
                        type: ERROR,
                        data: {
                            type,
                            data: error
                        },
                        meta: fromMiddleware({
                            ...meta
                        }),
                    }))

                    return next({
                        type: PROCESSING,
                        data: action,
                        meta: fromMiddleware({
                            promise
                        })
                    });

            } else {

                return next(action);

            }

        };

    }

}
