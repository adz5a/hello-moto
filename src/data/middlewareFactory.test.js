import {Â MiddlewareFactory } from "./middlewareFactory";
import isFunction from "lodash/isFunction";
import noop from "lodash/noop";
import {
    createStore,
    applyMiddleware
} from "redux";
import {
    ACTIONFACTORY
} from "data/commons";


const effect = x => x;
const actionType = "actionType";

const is = (expect, pred) => testee => expect(isFunction(testee)).toBe(true);

test("middleware factory", () => {

    const middleware = MiddlewareFactory({
        [actionType]: effect
    });
    const isFunc = is(expect, isFunction);

    // a middleware is a function
    isFunc(middleware);

    // a middleware returns :: ( next => ( action => action ) )
    isFunc(middleware());

    // should be a :: action => action 
    isFunc(middleware()());

});

describe("expect at least one argument", () => {

    test("must be a non null object", () => {

        // thrower
        [
            null,
            0,
            NaN,
            undefined,
            true,
            false
        ]
            .forEach(value => {

                expect( () => MiddlewareFactory(value) ).toThrowErrorMatchingSnapshot();

            });


        // does not throw
        [
            {}
        ]
            .forEach(value => {

                expect( () => MiddlewareFactory(value)).not.toThrow();

            });


    });

    test("each keys must be function", () => {


        // thrower
        [
            null,
            0,
            NaN,
            undefined,
            true,
            false
        ]
            .forEach(value => {

                expect( () => MiddlewareFactory({ key: value }) ).toThrowErrorMatchingSnapshot();

            });


        // does not throw
        [
            function () {},
            () => {}
        ]
            .forEach(value => {

                expect( () => MiddlewareFactory({ key: value }) ).not.toThrow();

            });

    });

});

describe("effects", () => {


    const prep = ( { effect = noop } = {} ) => {

        const reducer = ( state = {} ) => state;
        const scope = ACTIONFACTORY("test");
        const action = scope("action");
        const _effect = jest.fn(effect);
        const store = createStore(
            reducer, 
            applyMiddleware(MiddlewareFactory({ [action]: _effect }))
        );
        return {
            action,
            effect: _effect,
            reducer,
            store
        };


    };

    test("called once when action matches", () => {

        const { store, action, effect } = prep();

        store.dispatch({
            type: action
        });

        expect(effect).toHaveBeenCalledTimes(1);

        // return store.promise.then();
    });

    test("does not throw synchronously", () => {


        const { store, action, effect } = prep({ effect: () => {

            throw new Error("BOOM");

        } });

        expect(() => store.dispatch({
            type: action
        })).not.toThrow();

        expect(effect).toHaveBeenCalledTimes(1);
    });
    
});

