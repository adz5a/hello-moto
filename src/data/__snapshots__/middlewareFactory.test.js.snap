// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expect at least one argument each keys must be function 1`] = `"All keys must be functions, instead received non functions for : key"`;

exports[`expect at least one argument each keys must be function 2`] = `"All keys must be functions, instead received non functions for : key"`;

exports[`expect at least one argument each keys must be function 3`] = `"All keys must be functions, instead received non functions for : key"`;

exports[`expect at least one argument each keys must be function 4`] = `"All keys must be functions, instead received non functions for : key"`;

exports[`expect at least one argument each keys must be function 5`] = `"All keys must be functions, instead received non functions for : key"`;

exports[`expect at least one argument each keys must be function 6`] = `"All keys must be functions, instead received non functions for : key"`;

exports[`expect at least one argument must be a non null object 1`] = `"Expects an object :: { key: function }"`;

exports[`expect at least one argument must be a non null object 2`] = `"Expects an object :: { key: function }"`;

exports[`expect at least one argument must be a non null object 3`] = `"Expects an object :: { key: function }"`;

exports[`expect at least one argument must be a non null object 4`] = `"Expects an object :: { key: function }"`;

exports[`expect at least one argument must be a non null object 5`] = `"Expects an object :: { key: function }"`;

exports[`expect at least one argument must be a non null object 6`] = `"Expects an object :: { key: function }"`;
